// @udecode/plate-select/dist/index.es.js
import { isCollapsed, queryNode, createPluginFactory } from '@udecode/plate-core';
import { Editor, Transforms } from 'slate';

/**
 * Set a list of element types to select on backspace
 */
const withSelectOnBackspace = (editor, {
  options: {
    query
  }
}) => {
  const {
    deleteBackward
  } = editor;

  editor.deleteBackward = unit => {
    const {
      selection
    } = editor;

    if (unit === 'character' && isCollapsed(selection)) {
      const prevNode = Editor.before(editor, selection, {
        unit
      });

      if (prevNode) {
        const [prevCell] = Editor.nodes(editor, {
          match: node => queryNode([node, prevNode.path], query),
          at: prevNode
        });

        if (!!prevCell && prevNode) {
          // don't delete image, set selection there
          Transforms.select(editor, prevNode);
        } else {
          deleteBackward(unit);
        }
      } else {
        deleteBackward(unit);
      }
    } else {
      deleteBackward(unit);
    }
  };

  return editor;
};

const KEY_SELECT_ON_BACKSPACE = 'selectOnBackspace';
/**
 * @see {@link withSelectOnBackspace}
 */

const createSelectOnBackspacePlugin = createPluginFactory({
  key: KEY_SELECT_ON_BACKSPACE,
  withOverrides: withSelectOnBackspace
});

export { KEY_SELECT_ON_BACKSPACE, createSelectOnBackspacePlugin, withSelectOnBackspace };
