// @udecode/plate-line-height/dist/index.es.js
import { createPluginFactory, getPluginType, ELEMENT_DEFAULT, getPluginInjectProps, unsetNodes, setNodes } from '@udecode/plate-core';
import { Editor } from 'slate';

const KEY_LINE_HEIGHT = 'lineHeight';
/**
 * Enables support for text alignment, useful to align your content
 * to left, right and center it.
 */

const createLineHeightPlugin = createPluginFactory({
  key: KEY_LINE_HEIGHT,
  inject: {
    props: {
      nodeKey: KEY_LINE_HEIGHT,
      defaultNodeValue: 1.5
    }
  },
  then: editor => ({
    inject: {
      props: {
        validTypes: [getPluginType(editor, ELEMENT_DEFAULT)]
      }
    }
  })
});

const setLineHeight = (editor, {
  value,
  setNodesOptions
}) => {
  const {
    validTypes,
    defaultNodeValue,
    nodeKey
  } = getPluginInjectProps(editor, KEY_LINE_HEIGHT);

  const match = n => Editor.isBlock(editor, n) && !!validTypes && validTypes.includes(n.type);

  if (value === defaultNodeValue) {
    unsetNodes(editor, nodeKey, {
      match,
      ...setNodesOptions
    });
  } else {
    setNodes(editor, {
      [nodeKey]: value
    }, {
      match,
      ...setNodesOptions
    });
  }
};

export { KEY_LINE_HEIGHT, createLineHeightPlugin, setLineHeight };
