// @udecode/plate-line-height/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var plateCore = require('@udecode/plate-core');
var slate = require('slate');

const KEY_LINE_HEIGHT = 'lineHeight';
/**
 * Enables support for text alignment, useful to align your content
 * to left, right and center it.
 */

const createLineHeightPlugin = plateCore.createPluginFactory({
  key: KEY_LINE_HEIGHT,
  inject: {
    props: {
      nodeKey: KEY_LINE_HEIGHT,
      defaultNodeValue: 1.5
    }
  },
  then: editor => ({
    inject: {
      props: {
        validTypes: [plateCore.getPluginType(editor, plateCore.ELEMENT_DEFAULT)]
      }
    }
  })
});

const setLineHeight = (editor, {
  value,
  setNodesOptions
}) => {
  const {
    validTypes,
    defaultNodeValue,
    nodeKey
  } = plateCore.getPluginInjectProps(editor, KEY_LINE_HEIGHT);

  const match = n => slate.Editor.isBlock(editor, n) && !!validTypes && validTypes.includes(n.type);

  if (value === defaultNodeValue) {
    plateCore.unsetNodes(editor, nodeKey, {
      match,
      ...setNodesOptions
    });
  } else {
    plateCore.setNodes(editor, {
      [nodeKey]: value
    }, {
      match,
      ...setNodesOptions
    });
  }
};

exports.KEY_LINE_HEIGHT = KEY_LINE_HEIGHT;
exports.createLineHeightPlugin = createLineHeightPlugin;
exports.setLineHeight = setLineHeight;
