// @udecode/plate-line-height-ui/dist/index.es.js
import React, { useCallback } from 'react';
import { usePlateEditorState, getPluginInjectProps, isCollapsed, someNode } from '@udecode/plate-core';
import { KEY_LINE_HEIGHT, setLineHeight } from '@udecode/plate-line-height';
import { ToolbarDropdown, ToolbarButton } from '@udecode/plate-toolbar';
import { ReactEditor } from 'slate-react';

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

const LineHeightToolbarDropdown = props => {
  const [open, setOpen] = React.useState(false);
  const editor = usePlateEditorState();
  const {
    validNodeValues
  } = getPluginInjectProps(editor, KEY_LINE_HEIGHT);
  const onToggle = useCallback(() => {
    setOpen(!open);
  }, [open, setOpen]);
  const selectHandler = useCallback(lineHeight => {
    if (editor) {
      ReactEditor.focus(editor);
      setLineHeight(editor, {
        value: lineHeight
      });
    }
  }, [editor]);
  return /*#__PURE__*/React.createElement(ToolbarDropdown, {
    control: /*#__PURE__*/React.createElement(ToolbarButton, _extends({
      active: isCollapsed(editor === null || editor === void 0 ? void 0 : editor.selection) && someNode(editor, {
        match: n => n[KEY_LINE_HEIGHT] !== undefined
      })
    }, props)),
    open: open,
    onOpen: onToggle,
    onClose: onToggle
  }, validNodeValues && validNodeValues.map(lineHeight => /*#__PURE__*/React.createElement("div", {
    style: {
      cursor: 'pointer'
    },
    key: lineHeight,
    onClick: () => selectHandler(lineHeight)
  }, lineHeight)));
};

export { LineHeightToolbarDropdown };
