// @udecode/plate-alignment/dist/index.es.js
import { createPluginFactory, getPluginType, ELEMENT_DEFAULT, getPluginInjectProps, unsetNodes, setNodes } from '@udecode/plate-core';
import { Editor } from 'slate';

const KEY_ALIGN = 'align';
/**
 * Creates a plugin that adds alignment functionality to the editor.
 */

const createAlignPlugin = createPluginFactory({
  key: KEY_ALIGN,
  then: editor => ({
    inject: {
      props: {
        nodeKey: KEY_ALIGN,
        defaultNodeValue: 'left',
        styleKey: 'textAlign',
        validNodeValues: ['left', 'center', 'right', 'justify'],
        validTypes: [getPluginType(editor, ELEMENT_DEFAULT)]
      }
    }
  })
});

const setAlign = (editor, {
  key = KEY_ALIGN,
  value,
  setNodesOptions
}) => {
  const {
    validTypes,
    defaultNodeValue,
    nodeKey
  } = getPluginInjectProps(editor, key);

  const match = n => Editor.isBlock(editor, n) && !!validTypes && validTypes.includes(n.type);

  if (value === defaultNodeValue) {
    unsetNodes(editor, nodeKey, {
      match,
      ...setNodesOptions
    });
  } else {
    setNodes(editor, {
      [nodeKey]: value
    }, {
      match,
      ...setNodesOptions
    });
  }
};

export { KEY_ALIGN, createAlignPlugin, setAlign };
