// @udecode/plate-alignment/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var plateCore = require('@udecode/plate-core');
var slate = require('slate');

const KEY_ALIGN = 'align';
/**
 * Creates a plugin that adds alignment functionality to the editor.
 */

const createAlignPlugin = plateCore.createPluginFactory({
  key: KEY_ALIGN,
  then: editor => ({
    inject: {
      props: {
        nodeKey: KEY_ALIGN,
        defaultNodeValue: 'left',
        styleKey: 'textAlign',
        validNodeValues: ['left', 'center', 'right', 'justify'],
        validTypes: [plateCore.getPluginType(editor, plateCore.ELEMENT_DEFAULT)]
      }
    }
  })
});

const setAlign = (editor, {
  key = KEY_ALIGN,
  value,
  setNodesOptions
}) => {
  const {
    validTypes,
    defaultNodeValue,
    nodeKey
  } = plateCore.getPluginInjectProps(editor, key);

  const match = n => slate.Editor.isBlock(editor, n) && !!validTypes && validTypes.includes(n.type);

  if (value === defaultNodeValue) {
    plateCore.unsetNodes(editor, nodeKey, {
      match,
      ...setNodesOptions
    });
  } else {
    plateCore.setNodes(editor, {
      [nodeKey]: value
    }, {
      match,
      ...setNodesOptions
    });
  }
};

exports.KEY_ALIGN = KEY_ALIGN;
exports.createAlignPlugin = createAlignPlugin;
exports.setAlign = setAlign;
